pipeline/metadata/blockpage.py

```diff
@@ -1,55 +1,48 @@
 """Matcher for response pages to blockpage signatures."""
 
 import json
-import os
+import io
+import pkgutil
 import re
 from typing import Optional, Dict
 
-import apache_beam.io.filesystems as apache_filesystems
-
 # Signature filenames
-FALSE_POSITIVES = 'false_positive_signatures.json'
-BLOCKPAGES = 'blockpage_signatures.json'
+FALSE_POSITIVES = 'data/false_positive_signatures.json'
+BLOCKPAGES = 'data/blockpage_signatures.json'
 
 
-def _load_signatures(path: str) -> Dict[str, re.Pattern]:
+def _load_signatures(filepath: str) -> Dict[str, re.Pattern]:
   """Load signatures for blockpage matching.
 
   Args:
-    path: path to json file containing signatures
+    filepath: relative path to json file containing signatures
 
   Returns:
     Dictionary mapping fingerprints to signature patterns
   """
+  data = pkgutil.get_data(__name__, filepath)
+  if not data:
+    raise FileNotFoundError(f"Couldn't find file {filepath}")
+  content = io.TextIOWrapper(io.BytesIO(data), encoding='utf-8')
+
   signatures = {}
-  with apache_filesystems.FileSystems.open(path) as f:
-    for line in f:
-      if line != b'\n':
-        signature = json.loads(line.strip())
-        # Patterns stored in BigQuery syntax,
-        # so % represents any number of characters
-        pattern = signature['pattern']
-        # Convert to Python regex
-        pattern = re.escape(pattern)
-        pattern = pattern.replace('%', '.*')
-        signatures[signature['fingerprint']] = re.compile(pattern, re.DOTALL)
+  for line in content.readlines():
+    if line != '\n':
+      signature = json.loads(line.strip())
+      pattern = signature['pattern']
+      fingerprint = signature['fingerprint']
+
+      signatures[fingerprint] = re.compile(pattern, re.DOTALL)
   return signatures
 
 
 class BlockpageMatcher:
   """Matcher to confirm blockpages or false positives."""
 
-  def __init__(self, signature_folder: str) -> None:
-    """Create a Blockpage Matcher.
-    Args:
-      signature_folder: a folder containing signature files.
-        Either a gcs filepath or a local system folder.
-    """
-    false_positive_path = os.path.join(signature_folder, FALSE_POSITIVES)
-    blockpage_path = os.path.join(signature_folder, BLOCKPAGES)
-
-    self.false_positives = _load_signatures(false_positive_path)
-    self.blockpages = _load_signatures(blockpage_path)
+  def __init__(self) -> None:
+    """Create a Blockpage Matcher."""
+    self.false_positives = _load_signatures(FALSE_POSITIVES)
```

Comment:
Because the data is hardcoded, it makes it hard to test and reuse this class.

Instead:

- make constructor create an empty matcher 
- provide `add_blockpage(id, pattern)` and `add_false_positive(id, pattern)` methods
- provide a `create_default_matcher()` that reads from the default location and calls the add methods.

I'd recommend writing a pattern generator that you can reuse. Something like:
```py
def read_patterns(data):
  for line in data:
    # parse
    yield (id, pattern)
```
So you can have:
```py
for id, pattern in read_patterns(pkgutil.get_data(__name__, filepath)):
  matcher.add_blockpage(id, pattern)
```

Reply:
Strong disagree. As long as the data is inlined and inspectable I think this adds too much pattern overhead. The factory pattern is worth doing when we're reading data from an external cloud-project-level source, but it adds data-passing overhead which we already have a lot of.

I don't like the idea of having a constructor which creates an empty object which you have to do a bunch of extra setup to use. (Even if that setup can be encapsulated inside of a factory method.) It just adds complexity.

Tellingly, even if I made this change I don't think I would change any of the examples in `test_blockpage_matching` or `test_blockpage.py`. I'm already confident in the testing we're able to do for this class with the data inlined.

Reply:
What complexity are you referring to?
It doesn't add complexity to the client at all. Instead of calling BlockpageMatcher() you call create_default_blockpage_matcher().
Meanwhile, the tests are hard to read, since you need to know what patterns are being tested. I can't tell what's being tested without searching some other file. Furthermore, there's an unnecessary dependency on pkgutil and a hidden dependency on a global (the filepath) too deep in the code.

I just realized that this class has another problem: It does too much. It violates the single-responsibility principle.
Instead, you should have a simple PageMatcher() that all it does is take a page and return the pattern id. There's no reason for it to know about blockpages and false positivies. Then in the processing code you can have a matcher for blockpages and one for false positives. Rename `_load_signatures` to `create_matcher_from_file`.
This makes the concepts simpler and easier to understand, besides removing duplicate code and making the code more reusable and the tests more self-contained.

Because of the size of this PR, I'd be ok to fix this in a follow-up PR. But I'd like to see that change. We shouldn't mix low level and high level concepts. It makes understanding harder.

Reply:
added a TODO